amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -458,6 +466,90 @@ void runProcessLoop() {
         slog.resetInfoLogging();
     }
 
+    private boolean allParentShardLeasesDeleted(Lease lease, ShardInfo shardInfo) throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        boolean allParentShardLeasesDeleted = true;
+        for (String parentShard : lease.parentShardIds()) {
+            if (leaseCoordinator.leaseRefresher().getLease(ShardInfo.getLeaseKey(shardInfo, parentShard)) != null) {
+                allParentShardLeasesDeleted = false;
+            }
+        }
+        return allParentShardLeasesDeleted;
+    }
+
+    private void cleanupCompletedShards(Collection<ShardInfo> completedShards)
+            throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+        
+        for (ShardInfo completedShard : completedShards) {
+            final Lease lease = leaseCoordinator.getCurrentlyHeldLease(ShardInfo.leaseKeyProvider.apply(completedShard));
+            final Set<String> childShardKeys = lease.childShardIds().stream().map(c -> ShardInfo.getLeaseKey(completedShard, c)).collect(Collectors.toSet());
+
+            if (CollectionUtils.isNullOrEmpty(lease.childShardIds())) {
+                throw new IllegalStateException("No child shard state found in lease " + lease + " - not deleting lease.");
```

Comment:
How does it work during the rollforward upgrade ? Is this expected that all existing leases will have child shards ?

Reply:
This is covered by using a final getrecords call during shutdown now